jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/HttpChannelOverFCGI.java

```diff
@@ -35,114 +33,141 @@
-    private HttpInput.Content _specialContent;
+    private final Callback _asyncFillCallback = new AsyncFillCallback();
+    private final ServerFCGIConnection connection;
     private final HttpFields.Mutable fields = HttpFields.build();
     private final Dispatcher dispatcher;
+    private HttpInput.Content _specialContent;
     private String method;
     private String path;
     private String query;
     private String version;
     private HostPortHttpField hostPort;
 
-    public HttpChannelOverFCGI(Connector connector, HttpConfiguration configuration, EndPoint endPoint, HttpTransport transport)
+    public HttpChannelOverFCGI(ServerFCGIConnection connection, Connector connector, HttpConfiguration configuration, EndPoint endPoint, HttpTransport transport)
     {
         super(connector, configuration, endPoint, transport);
+        this.connection = connection;
         this.dispatcher = new Dispatcher(connector.getServer().getThreadPool(), this);
     }
 
     @Override
     public boolean onContent(HttpInput.Content content)
     {
-        boolean b = super.onContent(content);
+        boolean result = super.onContent(content);
 
-        Throwable failure;
-        try (AutoLock l = _lock.lock())
-        {
-            failure = _specialContent == null ? null : _specialContent.getError();
-            if (failure == null)
-                _contentQueue.offer(content);
-        }
-        if (failure != null)
+        Throwable failure = _specialContent == null ? null : _specialContent.getError();
+        if (failure == null)
+            _contentQueue.offer(content);
+        else
             content.failed(failure);
 
-        return b;
+        return result;
     }
 
     @Override
     public boolean needContent()
     {
-        try (AutoLock l = _lock.lock())
+        if (hasContent())
         {
-            boolean hasContent = _specialContent != null || !_contentQueue.isEmpty();
             if (LOG.isDebugEnabled())
-                LOG.debug("needContent has content? {}", hasContent);
-            return hasContent;
+                LOG.debug("needContent has immediate content {}", this);
+            return true;
         }
+
+        parseAndFill();
+
+        if (hasContent())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("needContent has parsed content {}", this);
+            return true;
+        }
+
+        connection.getEndPoint().tryFillInterested(_asyncFillCallback);
+        return false;
+    }
+
+    private boolean hasContent()
+    {
+        return _specialContent != null || !_contentQueue.isEmpty();
     }
 
     @Override
     public HttpInput.Content produceContent()
     {
-        HttpInput.Content content;
-        try (AutoLock l = _lock.lock())
+        if (!hasContent())
+            parseAndFill();
+
+        if (!hasContent())
+            return null;
+
+        HttpInput.Content content = _contentQueue.poll();
```

Comment:
Do we still need a queue here?  onContent now always returns true to ensure back pressure, so we should not get more than one content.
Also should we rename the action in the parse from ASYNC to IN_CONTENT?